packages/datetime/src/dateInput.tsx

```diff
@@ -164,7 +165,7 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
         );
 
         const inputClasses = classNames({
-            "pt-intent-danger": !(this.isMomentValidAndInRange(date) || this.isNull(date) || dateString === ""),
+            "pt-intent-danger": !(this.isMomentValidAndInRange(date) || isMomentNull(date) || dateString === ""),
```

Comment:
use the former. only use `intentClass()` if you don't know which intent it'll be.